zrml/swaps/rpc/src/lib.rs

```diff
@@ -205,4 +215,56 @@ where
             })
             .collect()
     }
+
+    async fn get_asset_spot_prices_for_pool(
+        &self,
+        pool_id: PoolId,
+        with_fees: bool,
+        blocks: Vec<NumberFor<Block>>,
+    ) -> RpcResult<BTreeMap<Asset<MarketId>, Vec<SerdeWrapper<Balance>>>> {
+        let api = self.client.runtime_api();
+        let at = BlockId::hash(self.client.info().best_hash);
+        let pool = api
+            .pool_by_id(&at, pool_id.clone())
+            .map_err(|e| {
+                CallError::Custom(ErrorObject::owned(
+                    Error::RuntimeError.into(),
+                    "Unable to get pool, Runtime trapped.",
+                    Some(e.to_string()),
+                ))
+            })?
+            .map_err(|e| {
+                CallError::Custom(ErrorObject::owned(
+                    Error::RuntimeError.into(),
+                    "Unable to get pool. DispatchError",
+                    Some(format!("{:?}", e)),
+                ))
+            })?;
+        let mut result_map = BTreeMap::<Asset<MarketId>, Vec<SerdeWrapper<Balance>>>::new();
+        for asset in &pool.assets {
```

Comment:
This seems like we're mixing business logic and RPC logic. The structure of the code should be the following:

- The runtime API exposes some of the storage/getter of `zrml_swap` to the RPC module
- The RPC module queries one of the entry points of the runtime API

In other words, you can expose a function `get_all_spot_prices` (or whatever its called) in `zrml_swaps`, which the runtime API exposes to the RPC module. This is the pattern that we've already used for all other RPC calls. The advantages (in addition to a clear separation of responsibility) are:

- The function below is much easier to write directly inside `zrml_swaps`.
- The behavior is testable without additional work. Just write the function in `zrml_swaps` and verify that it returns correct values there. Speaking of which, we definitely need tests for this function (can easily be derived from tests of `get_spot_price`).
- The RPC call is just a couple of lines long and looks pretty much exactly like the others.

Reply:
I think code won't be much cleaner than this, because we want prices of all assets over blocks specified in Vec<>. 
So if we add method in Swaps which just gives prices for all assets then also we need to iterate over all blocks and create vector and hasmap from the results. 
Or we can create BTreeMap with (asset, block_number) as key then it may be easy to collect().

Reply:
The code will definitely be cleaner (as mentioned above, separation of responsibilities and testability). I'd say we add `fn get_all_spot_prices(...) -> BTreeMap<Asset, Vec<Balance>>` to `zrml_swaps` and add tests. The runtime API directly exposes this functionality, and then all the RPC has to do something very similar to `get_spot_prices`.